core/store/src/trie/trie_recording.rs

```diff
@@ -7,22 +7,31 @@ use std::sync::Arc;
 /// A simple struct to capture a state proof as it's being accumulated.
 pub struct TrieRecorder {
     recorded: HashMap<CryptoHash, Arc<[u8]>>,
+    size: usize,
 }
 
 impl TrieRecorder {
     pub fn new() -> Self {
-        Self { recorded: HashMap::new() }
+        Self { recorded: HashMap::new(), size: 0 }
     }
 
     pub fn record(&mut self, hash: &CryptoHash, node: Arc<[u8]>) {
-        self.recorded.insert(*hash, node);
+        match self.recorded.insert(*hash, node.clone()) {
+            Some(old_node) => debug_assert_eq!(old_node, node, "Must record read-only nodes."),
+            None => self.size += node.len(),
+        }
     }
 
     pub fn recorded_storage(&mut self) -> PartialStorage {
         let mut nodes: Vec<_> = self.recorded.drain().map(|(_key, value)| value).collect();
         nodes.sort();
         PartialStorage { nodes: PartialState::TrieValues(nodes) }
     }
+
+    pub fn recorded_storage_size(&self) -> usize {
+        debug_assert!(self.size == self.recorded.values().map(|v| v.len()).sum::<usize>());
```

Comment:
This debug assert might be kinda expensive... If we are testing 1000 receipts and have 10000 recorded values then... might be kinda slow. What about we assert it at the very end of generating a state witness instead of every time?

Reply:
Conversely the debug assert expression gets evaluated in "development" builds only. While it is ideal to not have those be much slower than the release builds, there's already some precedent to debug asserts being slow enough to no longer give an accurate view of the release build performance, so if it is hard to move this elsewhere, then it is probably quite alright to leave it here...

Reply:
Yes, the reason why I had this as a part of debug_assert was because of the performance cost. This function is called once after every contract execution. Hopefully this shouldn't slow down things too much.